File: Game/src/main/java/net/winternetwork/bedwars/game/module/stage/StageModule.java
package net.winternetwork.bedwars.game.module.stage;

import lombok.Getter;
import net.winternetwork.bedwars.api.game.stage.Stage;
import net.winternetwork.bedwars.api.module.Module;
import net.winternetwork.bedwars.api.module.ModulePriority;
import net.winternetwork.bedwars.game.Game;
import net.winternetwork.bedwars.game.module.stage.flag.listener.FlagListener;

public class StageModule extends Module {

    private final Game game = Game.getGame();

    @Getter
    private StageManager stageManager;

    public StageModule() {
        super(
                "Stages",
                ModulePriority.HIGHEST
        );
    }

    @Override
    public void init() {
        log("Iniciando sistema de estágios...");
        this.stageManager = new StageManager();

        log("Registrando listeners...");
        Game.getGame()
                .registerListeners(new FlagListener(stageManager));
    }

    @Override
    public Runnable getScheduler() {
        return () -> {
            ++game.timeElapsed;

            final Stage stage = stageManager.getActualStage();

            stage.onSecondPassed();
            stage.operateTime();

            if (stage.getTimeLeft() <= 0) {
                stage.onTimeLeft();
            }
        };
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/game/modules/stage/StageModule.kt
package net.winternetwork.bedwars.game.modules.stage

import net.winternetwork.bedwars.api.module.Module
import net.winternetwork.bedwars.game.game
import net.winternetwork.bedwars.game.modules.maps.MapModule
import net.winternetwork.bedwars.game.modules.stage.`object`.GameStartedStage
import net.winternetwork.bedwars.game.modules.stage.`object`.ToStartStage
import net.winternetwork.bedwars.game.modules.stage.`object`.WaitingPlayerStage
import net.winternetwork.bedwars.game.modules.stage.listener.FlagListener
import net.winternetwork.bedwars.game.settings.GameSettings

class StageModule(mapModule: MapModule) : Module("Stages") {

    var stageManager: StageManager? = null
        private set

    private val stages = listOf(
            WaitingPlayerStage(this),
            ToStartStage(this, mapModule),
            GameStartedStage()
    )

    override fun init() {
        if (GameSettings.canStart) {
            log("Carregando estágios...")
            stageManager = StageManager(stages)

            log("Registrando listeners...")
            game.listeners(FlagListener(this))
        }
    }

    override fun onSecPassed() {
        if (!GameSettings.canStart) return

        stageManager?.actualStage?.run {
            onSecondPassed()
            operateTime()

            if (timeLeft <= 0)
                onTimeLeft()
        }
    }

}